library/old_protocol/old_protocol.cpp

```diff
@@ -0,0 +1,204 @@
+	switch (cmd.cmd) {
+	case DNET_CMD_LOOKUP_NEW:
+		enqueue_net(st, serialize_lookup_request(std::move(request)));
+		break;
+	default:
+		throw std::domain_error("old_protocol::send_request is not supported for "
+		                        "cmd = " + std::to_string(cmd.cmd));
+	}
+}
+
+std::unordered_set<int> old_protocol::supported_requests = { DNET_CMD_LOOKUP_NEW };
+std::unordered_set<int> old_protocol::supported_replies_on_srv = { DNET_CMD_LOOKUP_NEW };
+std::unordered_set<int> old_protocol::supported_replies_on_cl = {};
+
+int old_protocol::recv_message(dnet_cmd *cmd, dnet_net_state *st) {
+	if (!is_supported_message(cmd, st))
+		return -ENOTSUP;
+
+	dnet_logger_unset_trace_id();
+	dnet_logger_set_trace_id(cmd->trace_id, cmd->flags & DNET_FLAGS_TRACE_BIT);
+
+	data_pointer message_buffer;
+	int err = continue_read_message_after_cmd(cmd, st, message_buffer);
+	if (err)
+		return err;
+
+	if (cmd->flags & DNET_FLAGS_REPLY) {
+		auto it = repliers_table_.find(cmd->trans);
+		auto &repliers = it->second;
+
+		if (cmd->status) {
+			repliers.reply_error(cmd->status);
+
+		} else {
+			switch (cmd->cmd) {
+			case DNET_CMD_LOOKUP_NEW:
+				repliers.reply(deserialize_lookup_response(std::move(message_buffer)));
+				break;
+			default:
+				// Must never reach this code, due to is_supported_message() filter called before
+				return -ENOTSUP;
+			}
+		}
+
+		repliers_table_.erase(it);
+
+	} else {
+		switch (cmd->cmd) {
+		case DNET_CMD_LOOKUP_NEW:
+			lookup_request_translator_.translate_request(st, std::move(message_buffer));
+			break;
+		default:
+			// Must never reach this code, due to is_supported_message() filter called before
+			return -ENOTSUP;
+		}
+	}
+
+	// This is analogue of ending-part of pool.c/dnet_process_recv_single
+	clock_gettime(CLOCK_MONOTONIC_RAW, &st->rcv_finish_ts);
+
+	return 0;
+
+	// Some additional actions are done in caller: at pool.c/dnet_process_recv_single/label 'out'.
+}
+
+bool old_protocol::is_supported_message(dnet_cmd *cmd, dnet_net_state *st) {
+	auto &supported_cmds = [&]() -> std::unordered_set<int> & {
+		if (cmd->flags & DNET_FLAGS_REPLY) {
+			// For some message types we cannot yet accept replies addressed to elliptics-client, we can
+			// only accept replies from forward transactions (replies that are addressed to transitional
+			// server).
+			if (st->n2_tmp_forwarding_in_progress) {
+				if (!(cmd->flags & DNET_FLAGS_MORE))
+					st->n2_tmp_forwarding_in_progress = 0;
+
+				return supported_replies_on_srv;
+			} else
+				return supported_replies_on_cl;
+		} else
+			return supported_requests;
+	}();
+
+	return !!supported_cmds.count(cmd->cmd);
+}
+
+int old_protocol::continue_read_message_after_cmd(dnet_cmd *cmd,
+                                                  dnet_net_state *st,
+                                                  data_pointer &message_buffer) {
+	message_buffer = data_pointer::allocate(sizeof(dnet_cmd) + cmd->size);
+
+	memcpy(message_buffer.data(), cmd, sizeof(dnet_cmd));
```

Comment:
Let's not include `cmd` into `message_buffer`

Reply:
Done.